extensions/guacamole-auth-restrict/src/main/java/org/apache/guacamole/auth/restrict/user/RestrictedUserContext.java

```diff
@@ -0,0 +1,197 @@
+                return new RestrictedConnectionGroup(object, remoteAddress);
+            }
+
+            @Override
+            protected ConnectionGroup undecorate(ConnectionGroup object) {
+                assert(object instanceof RestrictedConnectionGroup);
+                return ((RestrictedConnectionGroup) object).getUndecorated();
+            }
+
+        };
+    }
+    
+    @Override
+    public Collection<Form> getConnectionGroupAttributes() {
+        Collection<Form> connectionGroupAttrs = new HashSet<>(super.getConnectionGroupAttributes());
+        connectionGroupAttrs.add(RestrictedConnectionGroup.RESTRICT_CONNECTIONGROUP_FORM);
+        return Collections.unmodifiableCollection(connectionGroupAttrs);
+    }
+    
+    @Override
+    public Directory<User> getUserDirectory() throws GuacamoleException {
+        return new DecoratingDirectory<User>(super.getUserDirectory()) {
+
+            @Override
+            protected User decorate(User object) {
+                return new RestrictedUser(object, remoteAddress);
+            }
+
+            @Override
+            protected User undecorate(User object) {
+                assert(object instanceof RestrictedUser);
+                return ((RestrictedUser) object).getUndecorated();
+            }
+
+        };
+    }
+    
+    @Override
+    public Collection<Form> getUserAttributes() {
+        Collection<Form> userAttrs = new HashSet<>(super.getUserAttributes());
+        userAttrs.add(RestrictedUser.RESTRICT_LOGIN_FORM);
+        return Collections.unmodifiableCollection(userAttrs);
+    }
+    
+    @Override
+    public Directory<UserGroup> getUserGroupDirectory() throws GuacamoleException {
+        return new DecoratingDirectory<UserGroup>(super.getUserGroupDirectory()) {
+            
+            @Override
+            protected UserGroup decorate(UserGroup object) {
+                return new RestrictedUserGroup(object);
+            }
+            
+            @Override
+            protected UserGroup undecorate(UserGroup object) {
+                assert(object instanceof RestrictedUserGroup);
+                return ((RestrictedUserGroup) object).getUndecorated();
+            }
+            
+        };
+    }
+    
+    @Override
+    public Collection<Form> getUserGroupAttributes() {
+        Collection<Form> userGroupAttrs = new HashSet<>(super.getUserGroupAttributes());
+        userGroupAttrs.add(RestrictedUserGroup.RESTRICT_LOGIN_FORM);
+        return Collections.unmodifiableCollection(userGroupAttrs);
+    }
+    
+    @Override
+    public boolean isValid() {
+        try {
+            // Verify whether or not time restrictions still apply.
+            RestrictionVerificationService.verifyTimeRestrictions(this, effectiveUserGroups);
+            return true;
+        }
+        catch (GuacamoleException e) {
+            LOGGER.debug("User account is now restricted and is no longer valid.");
```

Comment:
Maybe should include the exception here in the debug log?

Reply:
Yep, good idea. Done.